### DISCUSSION TITLE
Missing pulses

### DISCUSSION THREAD

**Discussion by macraj** [micropython/micropython#10518] [ESP8266]
Hello,
I am trying to make a gas counter. It is a simple code that reads pulses from a GPIO pin ( pulses are 4 seconds min to several hours max ), increments the counter and sends it periodically via MQTT. The problem is that it loses pulses (about 2%) for no reason. I checked the pulse readings on the Arduino, Uno are fine. The connection to the MQTT server is also ok, the wifi does not restart. I am puzzled how this code can not work. Could it be that Micropython or esp is doing something in the background and resetting the memory ?  But the counter is preserved except for a few missing pulses.  If anyone has any idea what could be the cause please give me a hint.

```py
def reconnect():
    while True:
        if not (wlan.ifconfig()[0] == '0.0.0.0'):
            try:
                mqtt_client.connect()
                break
            except Exception as ex:
                if debug:
                    print('error connecting MQTT server', ex)
                time.sleep(5)
    mqtt_client.publish(topic_id + "state/ip", wlan.ifconfig()[0])
    mem = str(gc.mem_free() // 4)
    mqtt_client.publish(topic_id + "state/mem", mem)
    mqtt_client.subscribe(topic_id + 'correct')


def sub_cb(topic, msg):
    global total_cnt
    # print("MQTT topic/message: {} / {}".format(topic, msg))
    total_cnt = float(msg)

    # mqtt_client.publish(topic_id + "total", str(total_cnt))
    if debug:
        print("MQTT change counter: {} / {}".format(msg, total_cnt))


def check_inp(reed_state):
    global reed_val, pulse, pin_reed
    led.value(not reed_state)
    if pulse:
        if not reed_state:
            reed_val += 1
            if reed_val > 2:
                reed_val = 0
                pulse = False
                return True
    else:
        if reed_state:
            reed_val += 1
        pulse = reed_val > 1


def main():
    global total_cnt, quarter_cnt, reed_val, pulse

    sync_nt()
    mqtt_client.set_callback(sub_cb)
    online = False
    sec = 0
    temp_cnt = 0
    while True:

        time.sleep(.1)
        reed_state = not pin_reed.value()
        if check_inp(reed_state):
            quarter_cnt += 1
            total_cnt += 0.01
            if debug:
                print('Input return: {} & {}'.format(quarter_cnt, total_cnt))
        if online:
            try:
                mqtt_client.check_msg()
                tm = time.localtime()
                if tm[5] != sec:  # every second
                    sec = tm[5]
                    if temp_cnt != total_cnt:
                        mqtt_client.publish(topic_id + "total", str(total_cnt))
                        temp_cnt = total_cnt
                    # if not (tm[5] % 15):
                    #     print('Pulse / reed {} # {} sec {}'.format(pulse, reed_val, sec))
                    #
                    if not sec:  # every minute
                        if debug:
                            print('Totalizer = {} / {}'.format(total_cnt, quarter_cnt))
                            # gc.collect()
                        mem = str(gc.mem_free() // 4)
                        mqtt_client.publish(topic_id + "state/mem", mem)
                        # mqtt_client.publish(topic_id + "state/online", str(time.time()-start_s))
                        if not (tm[4] % 15):
                            mqtt_client.publish(topic_id + "total", str(total_cnt))
                            mqtt_client.publish(topic_id + "quarter", str(quarter_cnt))
                            quarter_cnt = 0
            except Exception as ex:
                online = False
        else:
            reconnect()
            tm = time.localtime()
            r_tim = '{}-{:02d}-{:02d} {}:{:02d}'.format(tm[0], tm[1], tm[2], tm[3], tm[4])
            mqtt_client.publish(topic_id + "state/online", r_tim)
            online = True


main()

```

**Comment by robert-hh** [comment]
Just a note: Please enclose the whole code in lines with THREE backticks ```. Otherwise the code is not readable.

**Comment by 2dof** [comment]
Did you  try using interrupts, but from https://github.com/micropython/micropython/discussions/9759 it seems that missing counts still occur ,  in  https://github.com/micropython/micropython/issues/3775   there is some explanation , but maybe using both rising and falling edge could provide some redundancy. 

there is  machine counter implemented in miropython  https://micropython-tve.readthedocs.io/en/counter/library/machine.Counter.html  but seems not to be available for ESP82 (even on my esp32 ( mpy ver v1.19.1) is not available.

but  https://github.com/tve/mpy-lib/tree/master/esp32-counter   impelents (for esp32)  PCNT  maybe can be adapted.

**Comment by jimmo** [comment]
My best guess is that something in the networking code (either wifi management or mqtt) is blocking, preventing your pin checks from running.
As suggested above, try using a pin interrupt instead.

I'm also curious how you were able to confirm the missing pulses?

Another thing to consider would be to move the pin checking into its own dedicated thread.

**Reply by macraj** [reply]
Yes, I'm pretty sure it's something to do with the network. I've changed the code to deep sleep and every 15 minutes it wakes up, reads the values from the avr (via I2c) and then publishes them to MQTT. Of course I can also set the value ( counter data ) on the avr. So the avr is responsible for counting and storing the data.
How I'm able to confirm the missing pulses? Because it's normal analog gas meter, so I can read it from reels.
Now I'm trying  to play with esp32-c3, I'll see if it works better.

**Comment by peterhinch** [comment]
@macraj You have told us the pulse frequency, but what is its length?

@jimmo Assuming the official MQTT library the `connect` method calls `socket.getaddrinfo()` which is a blocking method. I don't think `publish` blocks unless `qos==1` in which case it blocks while waiting for `PUBACK` (which is a bug because `PUBACK` may never arrive).

**Reply by macraj** [reply]
Sorry for the late reply, I am now using an Arduino nano programmed in assembler and it works perfectly. No one pulse missed since 2 months. as for the pulses they vary from a few seconds to almost infinity. Occasionally the timer will stop on Pin On state for several hours, but shortest are no less then 4 sec. About  Esp it wakes up every 15 minutes, reads the Nano via I2C and publishes the data via mqtt also without problems.
